openmc/cell.py

```diff
@@ -0,0 +1,582 @@
+from collections import OrderedDict
+from collections.abc import Iterable
+from copy import deepcopy
+from math import cos, sin, pi
+from numbers import Real, Integral
+from xml.etree import ElementTree as ET
+import sys
+import warnings
+
+import numpy as np
+
+import openmc
+import openmc.checkvalue as cv
+from openmc.surface import Halfspace
+from openmc.region import Region, Intersection, Complement
+from openmc._xml import get_text
+from .mixin import IDManagerMixin
+
+
+class Cell(IDManagerMixin):
+    r"""A region of space defined as the intersection of half-space created by
+    quadric surfaces.
+
+    Parameters
+    ----------
+    cell_id : int, optional
+        Unique identifier for the cell. If not specified, an identifier will
+        automatically be assigned.
+    name : str, optional
+        Name of the cell. If not specified, the name is the empty string.
+    fill : openmc.Material or openmc.Universe or openmc.Lattice or None or iterable of openmc.Material, optional
```

Comment:
The casing logic that is done through out this file is based on this field.
We should enumerate `material`, `universe`, ... and define separate cell classes for each. That would make this much more maintainable.